### DISCUSSION TITLE
def VS var

### DISCUSSION THREAD

**Discussion by gsartori** [apache/grails-core#13842] [Polls]
> DISCLAIMER: I AM VOTING TO INTRODUCE `var` AS METHOD RETURN TYPE PLACEHOLDER IN GROOVY

At yesterday's meeting there was a quick discussion about using `var` instead of `def` for local variables to follow what Java is doing with its new `var` "keyword not keyword".

**I haven't slept tonight LOL (I'm laughing at myself, cause that statement is true 😆).**

These are my nightly thoughts:
- Groovy uses `def` to define variables and as return type on methods. `def` in Groovy is an alias for `Object`.
- Python uses `def` to define variables and methods. Python lately become well known due to AI so I kind of guess that `def` is not so alien anymore to developers (but yes, that's just speculation).
- Java > 10 intriduced `var` with type inference on assignment. That triggered different opinions on Java developers, not all of them like the new feature.
- Groovy introduced `var` as an alternative to define variables to stay familiar with Java developers. **It does not use `var` as a return type in methods.**
- Javascript introduced `let` to fix the scope issue they have with `var`. That triggered different opinions on JavaScript developers, not all of them like the new feature. They understand the reason why, but they keep using `var` cause in the end it's more expressive and they don't care much about scoping variables (that's not what I dediced to do with Dueuno Elements but that's my personal opinion based on random talks with my friends).
- **Complicating is easy, simplifying is complicated**

I loved Python and at the same time I love the nature of Groovy of being familiar to java developers. So I like both `def` and `var`. I would be the first one to find/replace `def` -> `var` in Dueuno Elements if that would work out of the box. But it doesn't.

**What I don't like is using both of them: `var` for variables and `def` for methods (like in Grails controllers !!).**

**Comment by erickinsella** [comment]
For context, I believe, the discussion was mainly around the code in grails-core (and supported plugins). Going forward as committers to those repos, should we favor using var or def. I believe there are good points to be made for both usages. I don't remember there being talk of Groovy changing anything. So, I think, the poll above might be a bit misleading. Recently, similar chatter has come up about code formatting and standardization of styles but nothing has been committed to thus far. I look forward to more discussion here with the wider grails community.

**Reply by gsartori** [reply]
Thank you @erickinsella, yes I have broaden the discussion starting from what I've heard at the meeting. I may be wrong but I think it is worth talking about it. I don't know why the Groovy engineers decided to have a `var` alias for local variables but not for method return types. Since the `def` keyword has been documented as a short for `Object` that's the way I see it in method declarations as well. Maybe someone involved can say something about it.

**Reply by gsartori** [reply]
@paulk-asert, do you have any words of wisdom?

**Reply by paulk-asert** [reply]
The "var" reserved type was introduced in Java for type inference of local **var**iables. Groovy added support mostly for Java-compatibility reasons. It's a Groovy feature that you can cut-n-paste a lot of Java code and it will still work with Groovy. But since Java doesn't support var for method definitions, the motivation for that change would be different. Groovy did extend var usage to allow it also to be used for fields/properties (which some considered a little controversial) which are similar in form to local variable definitions but method definitions would be a bigger difference to Java and we'd want a stronger justification as to why we'd support that.

**Reply by gsartori** [reply]
@paulk-asert if we want Groovy to be Java compatible (meaning that a Java program would be compiled by Groovy) having a `var` return type in Groovy would be fine, since the goal is not to make a Groovy progam be compiled by Java.

Groovy started as a _Python-ish_ language and is becoming more and more _Java-ish_. So if we have `var` now, why should we write python-ish code only when implementing methods?

**Comment by codeconsole** [comment]
This really doesn't even remotely represent the original discussion.   

There was no proposal to change how `var` works nor would we ever want to change how `var` works.  Not a single person mentioned asking the Groovy team to change anything.

**Reply by gsartori** [reply]
@codeconsole where exactly in my post I am stating that someone "mentioned asking the Groovy team to change anything"? That is something I have deliberately added to the poll.

**Reply by jeffscottbrown** [reply]
> There was no proposal to change how var works nor would we ever want to change how var works. Not a single person mentioned asking the Groovy team to change anything.


I think that is all true, but I don't think the original post above is suggesting otherwise.  I took those notes to be @gsartori sharing their own personal initial thoughts. "These are my nightly thoughts:"

**Reply by gsartori** [reply]
@jeffscottbrown that was indeed the purpose of the post

**Comment by codeconsole** [comment]
What is the point of the poll? to make a subjective decision?  You are forcing people to agree with reasoning by making a choice.   I would not vote for either of those options as I don't agree with either of those statements.  I recommend  if you are going to conduct a poll, you provide options without forcing reasoning or personal opinion behind them.  You should also provide options that account for all possible scenarios. Like only use `var` for variables in documentation examples so that people coming from Java background would better understand what is going on.

Books and old documentation is not a reason for keeping the status quo.  Should I not use `var` in Java then? Should I be using xml with Spring because there is a plethora of books and documentation teaching how to configure beans with Spring?  Should I not use interfaces with default methods in Java and stick with abstract classes?  Languages change and evolve.  If it was up to me, I would eliminate the semi-colon from Java. 😮 

The whole point of the original discussion in the meeting was writing code in a manner that is best understood to the widest audience.  At no point did anyone say go do a search and replace for every `def` statement with `var`.  The discussion was based around the examples provided by the Groovy team for migration from [the legacy date classes.](https://groovy.apache.org/blog/groovy-dates-and-times-cheat).

Don't get me wrong, I appreciate your involvement and the passion for the language. We just need to be mindful on what the best pathway forward is to attracting new involvement.  This wouldn't even be a discussion it there was an overwhelming number of people contributing to Grails, but there isn't.  I am **one** of a _few_ number of  people actively contributing and that needs to change. 

 If you want to submit PRs that utilize `def` to the Grails codebases, they would happily be accepted. We would love the help!

**Reply by jeffscottbrown** [reply]
"If you want to submit PRs that utilize def to the Grails codebases, they would happily be accepted." - Is that to suggest that `var` would not be accepted?

**Reply by codeconsole** [reply]
no, `var` would happily be accepted as well 😉  
The more contribution, the better 😄

**Reply by gsartori** [reply]
@codeconsole I was not representing any discussion about `var` I was just triggered by that discussion (the same way you've been triggered by my post 😉) and I wrote my point of view witch is subjective, of course.

The poll is not there to take decisions, how can it be? I am nobody to force anybody to do anything. The poll is there to see what people think and yes it's polarizing. I wanted to make it that way to trigger a discussion since we have a problem here which is not "`var` is better that `def` from a Java developer perspective" (which I agree, btw) but "The Groovy `var` is not a replacement for `def` since it only works with variables".

I would like to use `var` but what I see is a semantic pirfall that forces us to use BOTH `def` AND `var` if we want to use `var`.

**If `def` is an alias for `Object` and `var` is an alias for `def` then `var` must be an alias for `Object` but that is not the case.**

I can understand that the meaning of the word 'variable' does not fit well with the definition of a method ('def' meaning "define" i guess is a more general term so it fits better) but in Groovy `def` stands for `Object` so for example (Tag: @paulk-asert):

```
def index() {
   ...
}
```

Means:

```
Object index() {
   ...
}
```

Which, with the introduciotn of `var` as an alias for `def`, should let us write:

```
var index() {
   ...
}
```

Now, is it bad? Is it good? There are reasons for both. My very personal opinion is that if Groovy wants to be closer to Java than closer to Python, well than we should be able to use `var` for methods too and be able to write code without any `def`.

Because in Groovy `def` represents "any" return type, it is not a keyword to define a method like in Python. So in our case `var` means "this method returns a variable, if you don't want to return a variable use `void`".

At least this is my understanding from the documentation, and I am very happy to be proven wrong.

**Reply by gsartori** [reply]
@paulk-asert

**Reply by codeconsole** [reply]
@gsartori `var` is not an alias for `def`, nor was it being described or recommended to be used as one.   The discussion was using it how it is supported currently in Java and only that way. We never said make it an alias for anything. 

```
var index() {
   ...
}
```
would never make any sense to use it in such a way.   Did you even [look at what was referred to](https://groovy.apache.org/blog/groovy-dates-and-times-cheat) in the meeting?  Those examples make sense and would make sense to a Java developer.  Your example would not.

We are only talking about using it in situations where the compiler knows how it will be used at compile time:
```
var s = new String()
var i = 100
var c = new Car()
```

Illegal usage
```
var s = null 
```
[Read more about proper use of var here](https://www.baeldung.com/java-10-local-variable-type-inference)


It would never be used to replace Object. lt is used for static typing, not dynamic typing.  Using it in a way other than how it is used in Java would not serve the purpose of the original discussion and be extremely confusing. The point of the discussion was to attract more Java developers, not confuse them.

**Reply by gsartori** [reply]
@codeconsole what you just wrote about using `var` is fine with me and **is not the subject of my post.**

Let me underline this once and for all:

The
subject
of
my
post
is
not
directly
related
to
the
discussion
you
are
mentioning.

That discussion just triggered me. Full stop. It ends there.
In my post I am talking about something completely different.

From the Groovy documentation: https://groovy-lang.org/semantics.html#_variable_definition

> def and var act as a type placeholder, i.e. a replacement for the type name, when you do not want to give an explicit type. It could be that you don’t care about the type at compile time or are relying on type inference (with Groovy’s static nature). It is mandatory for variable definitions to have a type or placeholder. If left out, the type name will be deemed to refer to an existing variable (presumably declared earlier). For scripts, undeclared variables are assumed to come from the Script binding. In other cases, you will get a missing property (dynamic Groovy) or compile time error (static Groovy). **If you think of def and var as an alias of Object, you will understand in an instant.**

https://groovy-lang.org/semantics.html#_optional_typing

> [..] def is, in Groovy, strictly equivalent to using Object

> [..] Eventually, the type can be removed altogether from both the return type and the descriptor. But if you want to remove it from the return type, you then need to add an explicit modifier for the method, so that the compiler can make a difference between a method declaration and a method call

Given the above pargraphs we should be able to use `var` as return type placeholder as well.
> DISCLAIMER: I AM VOTING TO INTRODUCE var AS METHOD RETURN TYPE PLACEHOLDER IN GROOVY

My sub-optimal brain doesn't like using `def` and `var` in different ways in different contexts. Coherence always works and exceptions comes with a cost. Always. If `def` and `var` are the same thing they should be fully interchangable (Tag: @paulk-asert).

**Hence the poll and my PERSONAL AND SUBJECTIVE opinion that I wanted to share: As a general guideline, should we keep coding using `def` or should we be able to use `var` in methods as well?**

**Comment by gsartori** [comment]
**A summary from ChatGpt:**

The discussion in this GitHub thread revolves around the potential for Groovy to support using `var` as a method return type, similar to how it works for local variables. Here's a summary of the key points:

- **Introduction to the Idea:** The author, gsartori, starts by sharing their personal thoughts on the subject. They propose that `var` should be allowed as a return type for methods in Groovy, similar to its use for local variables. They suggest this change could improve consistency between Groovy and Java, which recently adopted `var` for type inference in local variables (Java 10+).

- **Current Use in Groovy:** In Groovy, `def` is used both for defining local variables and as a return type for methods. `def` is essentially an alias for `Object` in method return types, and Groovy has added `var` for local variables for Java compatibility. However, `var` cannot currently be used as a method return type in Groovy, unlike Java where `var` is only used for local variables.

**Community Response:** Several contributors weigh in:

- **erickinsella** clarifies that the discussion is more about which keyword (`def` or `var`) to use in Grails codebases and does not suggest changes to Groovy itself.

- **paulk-asert** explains that while Java introduced `var` for local variable type inference, it doesn't make sense to introduce `var` as a return type in Groovy because it would confuse the language's dynamic nature and diverge from Java compatibility.

- **codeconsole** adds that the goal is not to replace `def` with `var` everywhere, but rather to use `var` in ways that align with Java's intended use, ensuring better understanding for Java developers.

**The Poll:** gsartori creates a poll to gather community opinions, asking whether Groovy should promote `var` as a return type for methods, which would make `var` interchangeable with `def` in all contexts. However, the poll is intended as a discussion starter rather than a decision-making tool.

**Main Points of Contention:**

- Some believe Groovy should remain Java-compatible and use `var` only for local variables.

- Others feel that for consistency, `var` should be allowed as a return type for methods, similar to how `def` works.

- There is a concern that using both `def` and `var` in different contexts creates confusion and inconsistency.

In conclusion, the conversation is about whether Groovy should allow `var` as a method return type, making it fully interchangeable with `def`, and whether such a change would better align with Java or Python conventions. The poll is a way to gauge community opinion but is framed as a subjective reflection of gsartori's views.
